### ISSUE TITLE
how i can get recall for every class when iou_thr = 0.5

### ISSUE THREAD

**Issue by THEBEST-cloud** [open-mmlab/mmdetection#10921]
### env_info
---------------------------------------------------------------------------------------------------------------------------------------------------
sys.platform: win32
Python: 3.8.16 (default, Jun 12 2023, 21:00:42) [MSC v.1916 64 bit (AMD64)]
CUDA available: True
numpy_random_seed: 2147483648
GPU 0: NVIDIA GeForce RTX 3060 Ti
CUDA_HOME: None
MSVC: 用于 x64 的 Microsoft (R) C/C++ 优化编译器 19.36.32535 版
GCC: n/a
PyTorch: 1.8.0
PyTorch compiling details: PyTorch built with:
  - C++ Version: 199711
  - MSVC 192829337
  - Intel(R) Math Kernel Library Version 2020.0.0 Product Build 20191125 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.7.0 (Git Hash 7aed236906b1f7a05c0917e5257a1af05e9ff683)
  - OpenMP 2019
  - CPU capability usage: AVX2
  - CUDA Runtime 11.1
  - NVCC architecture flags: -gencode;arch=compute_37,code=sm_37;-gencode;arch=compute_50,code=sm_50;-gencode;arch=compute_60,code=sm_60;-gencode;arch=compute_61,code=sm_61;-gencode;arch=compute_70,code=sm_70;-gencode;arch=compute_75,code=sm_75;-gencode;arch=compute_80,code=sm_80;-gencode;arch=compute_86,code=sm_86;-gencode;arch=compute_37,code=compute_37
  - CuDNN 8.0.5
  - Magma 2.5.4
  - Build settings: BLAS_INFO=mkl, BUILD_TYPE=Release, CUDA_VERSION=11.1, CUDNN_VERSION=8.0.5, CXX_COMPILER=C:/cb/pytorch_1000000000000/work/tmp_bin/sccache-cl.exe, CXX_FLAGS=/DWIN32 /D_WINDOWS /GR /EHsc /w /bigobj -DUSE_PTHREADPOOL -openmp:experimental -DNDEBUG -DUSE_FBGEMM -DUSE_XNNPACK, LAPACK_INFO=mkl, PERF_WITH_AVX=1, PERF_WITH_AVX2=1, PERF_WITH_AVX512=1, TORCH_VERSION=1.8.0, USE_CUDA=ON, USE_CUDNN=ON, USE_EXCEPTION_PTR=1, USE_GFLAGS=OFF, USE_GLOG=OFF, USE_MKL=ON, USE_MKLDNN=ON, USE_MPI=OFF, USE_NCCL=OFF, USE_NNPACK=OFF, USE_OPENMP=ON, 

TorchVision: 0.9.0
OpenCV: 4.8.0
MMEngine: 0.8.0
mmdetection: 3.1.0+f78af77
------------------------------------------------------------------------------------------------------------------------------------------------
### config
train_cfg = dict(type='EpochBasedTrainLoop', max_epochs=40, val_interval=50)
val_cfg = dict(type='ValLoop')
test_cfg = dict(type='TestLoop')
param_scheduler = [
    dict(
        type='LinearLR', start_factor=0.0001, by_epoch=False, begin=0, end=70),
    dict(
        type='CosineAnnealingLR',
        eta_min=3.125e-05,
        begin=20,
        end=40,
        T_max=20,
        by_epoch=True,
        convert_to_iter_based=True),
]
optim_wrapper = dict(
    type='OptimWrapper',
    optimizer=dict(
        type='SGD',
        lr=0.000625,
        momentum=0.9,
        weight_decay=0.0005,
        nesterov=True),
    paramwise_cfg=dict(norm_decay_mult=0.0, bias_decay_mult=0.0))
auto_scale_lr = dict(enable=False, base_batch_size=64)
default_scope = 'mmdet'
default_hooks = dict(
    timer=dict(type='IterTimerHook'),
    logger=dict(type='LoggerHook', interval=50),
    param_scheduler=dict(type='ParamSchedulerHook'),
    checkpoint=dict(type='CheckpointHook', interval=50, max_keep_ckpts=3),
    sampler_seed=dict(type='DistSamplerSeedHook'),
    visualization=dict(type='DetVisualizationHook'))
env_cfg = dict(
    cudnn_benchmark=False,
    mp_cfg=dict(mp_start_method='fork', opencv_num_threads=0),
    dist_cfg=dict(backend='nccl'))
vis_backends = [
    dict(type='LocalVisBackend'),
]
visualizer = dict(
    type='DetLocalVisualizer',
    vis_backends=[
        dict(type='LocalVisBackend'),
    ],
    name='visualizer')
log_processor = dict(type='LogProcessor', window_size=50, by_epoch=True)
log_level = 'INFO'
load_from = None
resume = True
tta_model = dict(
    type='DetTTAModel',
    tta_cfg=dict(nms=dict(type='nms', iou_threshold=0.65), max_per_img=100))
img_scales = [
    (
        640,
        640,
    ),
    (
        320,
        320,
    ),
    (
        960,
        960,
    ),
]
tta_pipeline = [
    dict(type='LoadImageFromFile', backend_args=None),
    dict(
        type='TestTimeAug',
        transforms=[
            [
                dict(type='Resize', scale=(
                    640,
                    640,
                ), keep_ratio=True),
                dict(type='Resize', scale=(
                    320,
                    320,
                ), keep_ratio=True),
                dict(type='Resize', scale=(
                    960,
                    960,
                ), keep_ratio=True),
            ],
            [
                dict(type='RandomFlip', prob=1.0),
                dict(type='RandomFlip', prob=0.0),
            ],
            [
                dict(
                    type='Pad',
                    pad_to_square=True,
                    pad_val=dict(img=(
                        114.0,
                        114.0,
                        114.0,
                    ))),
            ],
            [
                dict(type='LoadAnnotations', with_bbox=True),
            ],
            [
                dict(
                    type='PackDetInputs',
                    meta_keys=(
                        'img_id',
                        'img_path',
                        'ori_shape',
                        'img_shape',
                        'scale_factor',
                        'flip',
                        'flip_direction',
                    )),
            ],
        ]),
]
img_scale = (
    640,
    640,
)
model = dict(
    type='YOLOX',
    data_preprocessor=dict(
        type='DetDataPreprocessor',
        pad_size_divisor=32,
        batch_augments=[
            dict(
                type='BatchSyncRandomResize',
                random_size_range=(
                    480,
                    800,
                ),
                size_divisor=32,
                interval=10),
        ]),
    backbone=dict(
        type='CSPDarknet',
        deepen_factor=0.33,
        widen_factor=0.5,
        out_indices=(
            2,
            3,
            4,
        ),
        use_depthwise=False,
        spp_kernal_sizes=(
            5,
            9,
            13,
        ),
        norm_cfg=dict(type='BN', momentum=0.03, eps=0.001),
        act_cfg=dict(type='Swish')),
    neck=dict(
        type='YOLOXPAFPN',
        in_channels=[
            128,
            256,
            512,
        ],
        out_channels=128,
        num_csp_blocks=1,
        use_depthwise=False,
        upsample_cfg=dict(scale_factor=2, mode='nearest'),
        norm_cfg=dict(type='BN', momentum=0.03, eps=0.001),
        act_cfg=dict(type='Swish')),
    bbox_head=dict(
        type='YOLOXHead',
        num_classes=9,
        in_channels=128,
        feat_channels=128,
        stacked_convs=2,
        strides=(
            8,
            16,
            32,
        ),
        use_depthwise=False,
        norm_cfg=dict(type='BN', momentum=0.03, eps=0.001),
        act_cfg=dict(type='Swish'),
        loss_cls=dict(
            type='CrossEntropyLoss',
            use_sigmoid=True,
            reduction='sum',
            loss_weight=1.0),
        loss_bbox=dict(
            type='IoULoss',
            mode='square',
            eps=1e-16,
            reduction='sum',
            loss_weight=5.0),
        loss_obj=dict(
            type='CrossEntropyLoss',
            use_sigmoid=True,
            reduction='sum',
            loss_weight=1.0),
        loss_l1=dict(type='L1Loss', reduction='sum', loss_weight=1.0)),
    train_cfg=dict(assigner=dict(type='SimOTAAssigner', center_radius=2.5)),
    test_cfg=dict(score_thr=0.01, nms=dict(type='nms', iou_threshold=0.65)))
data_root = 'E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO'
dataset_type = 'CocoDataset'
backend_args = None
train_pipeline = [
    dict(type='Mosaic', img_scale=(
        640,
        640,
    ), pad_val=114.0),
    dict(
        type='RandomAffine',
        scaling_ratio_range=(
            0.1,
            2,
        ),
        border=(
            -320,
            -320,
        )),
    dict(
        type='MixUp',
        img_scale=(
            640,
            640,
        ),
        ratio_range=(
            0.8,
            1.6,
        ),
        pad_val=114.0),
    dict(type='YOLOXHSVRandomAug'),
    dict(type='RandomFlip', prob=0.5),
    dict(type='Resize', scale=(
        640,
        640,
    ), keep_ratio=True),
    dict(
        type='Pad',
        pad_to_square=True,
        pad_val=dict(img=(
            114.0,
            114.0,
            114.0,
        ))),
    dict(type='FilterAnnotations', min_gt_bbox_wh=(
        1,
        1,
    ), keep_empty=False),
    dict(type='PackDetInputs'),
]
train_dataset = dict(
    type='MultiImageMixDataset',
    dataset=dict(
        type='CocoDataset',
        data_root='data/coco/',
        ann_file='annotations/instances_train2017.json',
        data_prefix=dict(img='train2017/'),
        pipeline=[
            dict(type='LoadImageFromFile', backend_args=None),
            dict(type='LoadAnnotations', with_bbox=True),
        ],
        filter_cfg=dict(filter_empty_gt=False, min_size=32),
        backend_args=None),
    pipeline=[
        dict(type='Mosaic', img_scale=(
            640,
            640,
        ), pad_val=114.0),
        dict(
            type='RandomAffine',
            scaling_ratio_range=(
                0.1,
                2,
            ),
            border=(
                -320,
                -320,
            )),
        dict(
            type='MixUp',
            img_scale=(
                640,
                640,
            ),
            ratio_range=(
                0.8,
                1.6,
            ),
            pad_val=114.0),
        dict(type='YOLOXHSVRandomAug'),
        dict(type='RandomFlip', prob=0.5),
        dict(type='Resize', scale=(
            640,
            640,
        ), keep_ratio=True),
        dict(
            type='Pad',
            pad_to_square=True,
            pad_val=dict(img=(
                114.0,
                114.0,
                114.0,
            ))),
        dict(
            type='FilterAnnotations',
            min_gt_bbox_wh=(
                1,
                1,
            ),
            keep_empty=False),
        dict(type='PackDetInputs'),
    ])
test_pipeline = [
    dict(type='LoadImageFromFile', backend_args=None),
    dict(type='Resize', scale=(
        640,
        640,
    ), keep_ratio=True),
    dict(
        type='Pad',
        pad_to_square=True,
        pad_val=dict(img=(
            114.0,
            114.0,
            114.0,
        ))),
    dict(type='LoadAnnotations', with_bbox=True),
    dict(
        type='PackDetInputs',
        meta_keys=(
            'img_id',
            'img_path',
            'ori_shape',
            'img_shape',
            'scale_factor',
        )),
]
train_dataloader = dict(
    batch_size=4,
    num_workers=4,
    persistent_workers=True,
    sampler=dict(type='DefaultSampler', shuffle=True),
    dataset=dict(
        type='MultiImageMixDataset',
        dataset=dict(
            type='CocoDataset',
            data_root='E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO',
            ann_file=
            'E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO/train/annotations/train.json',
            data_prefix=dict(
                img='E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO/train/images/'
            ),
            pipeline=[
                dict(type='LoadImageFromFile', backend_args=None),
                dict(type='LoadAnnotations', with_bbox=True),
            ],
            filter_cfg=dict(filter_empty_gt=False, min_size=32),
            backend_args=None,
            metainfo=dict(classes=('DT_VOID', ), palette=[
                (
                    255,
                    87,
                    34,
                ),
            ])),
        pipeline=[
            dict(type='Mosaic', img_scale=(
                640,
                640,
            ), pad_val=114.0),
            dict(
                type='RandomAffine',
                scaling_ratio_range=(
                    0.1,
                    2,
                ),
                border=(
                    -320,
                    -320,
                )),
            dict(
                type='MixUp',
                img_scale=(
                    640,
                    640,
                ),
                ratio_range=(
                    0.8,
                    1.6,
                ),
                pad_val=114.0),
            dict(type='YOLOXHSVRandomAug'),
            dict(type='RandomFlip', prob=0.5),
            dict(type='Resize', scale=(
                640,
                640,
            ), keep_ratio=True),
            dict(
                type='Pad',
                pad_to_square=True,
                pad_val=dict(img=(
                    114.0,
                    114.0,
                    114.0,
                ))),
            dict(
                type='FilterAnnotations',
                min_gt_bbox_wh=(
                    1,
                    1,
                ),
                keep_empty=False),
            dict(type='PackDetInputs'),
        ]),
    pin_memory=False)
val_dataloader = dict(
    batch_size=1,
    num_workers=1,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='CocoDataset',
        data_root='E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO',
        ann_file=
        'E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO/test/annotations/test.json',
        data_prefix=dict(
            img='E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO/test/images/'),
        test_mode=True,
        pipeline=[
            dict(type='LoadImageFromFile', backend_args=None),
            dict(type='Resize', scale=(
                640,
                640,
            ), keep_ratio=True),
            dict(
                type='Pad',
                pad_to_square=True,
                pad_val=dict(img=(
                    114.0,
                    114.0,
                    114.0,
                ))),
            dict(type='LoadAnnotations', with_bbox=True),
            dict(
                type='PackDetInputs',
                meta_keys=(
                    'img_id',
                    'img_path',
                    'ori_shape',
                    'img_shape',
                    'scale_factor',
                )),
        ],
        backend_args=None,
        metainfo=dict(classes=('DT_VOID', ), palette=[
            (
                255,
                87,
                34,
            ),
        ])),
    pin_memory=False)
test_dataloader = dict(
    batch_size=1,
    num_workers=1,
    persistent_workers=True,
    drop_last=False,
    sampler=dict(type='DefaultSampler', shuffle=False),
    dataset=dict(
        type='CocoDataset',
        data_root='E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO',
        ann_file=
        'E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO/test/annotations/test.json',
        data_prefix=dict(
            img='E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO/test/images/'),
        test_mode=True,
        pipeline=[
            dict(type='LoadImageFromFile', backend_args=None),
            dict(type='Resize', scale=(
                640,
                640,
            ), keep_ratio=True),
            dict(
                type='Pad',
                pad_to_square=True,
                pad_val=dict(img=(
                    114.0,
                    114.0,
                    114.0,
                ))),
            dict(type='LoadAnnotations', with_bbox=True),
            dict(
                type='PackDetInputs',
                meta_keys=(
                    'img_id',
                    'img_path',
                    'ori_shape',
                    'img_shape',
                    'scale_factor',
                )),
        ],
        backend_args=None,
        metainfo=dict(classes=('DT_VOID', ), palette=[
            (
                255,
                87,
                34,
            ),
        ])),
    pin_memory=False)
val_evaluator = dict(
    type='CocoMetric',
    ann_file=
    'E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO/test/annotations/test.json',
    metric='bbox',
    backend_args=None,
    classwise=True)
test_evaluator = dict(
    type='CocoMetric',
    ann_file=
    'E:\\lijunjie\\GPR\\radar_data\\AUD_VOID_COCO/test/annotations/test.json',
    metric='bbox',
    backend_args=None,
    classwise=True)
max_epochs = 40
num_last_epochs = 15
interval = 10
base_lr = 0.000625
custom_hooks = [
    dict(type='YOLOXModeSwitchHook', num_last_epochs=15, priority=48),
    dict(type='SyncNormHook', priority=48),
    dict(
        type='EMAHook',
        ema_type='ExpMomentumEMA',
        momentum=0.0001,
        update_buffers=True,
        priority=49),
]
num_classes = 1
metainfo = dict(
    classes=('DT_VOID', ), palette=[
        (
            255,
            87,
            34,
        ),
    ])
train_batch_size_per_gpu = 4
trian_num_workers = 4
val_batch_size_per_gpu = 1
val_num_workers = 1
resume_from = 'configs\\GPR\\yolox_s_8x8_300e_coco_20211121_095711-4592a793.pth'
work_dir = '.\\work_dirs\\yolox_s_8xb8-300e_void'
